src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt

```diff
@@ -1044,7 +1044,11 @@ class RsTypeInferenceWalker(
             blockTys.add(elseBranch.ifExpr?.inferType(expected))
             blockTys.add(elseBranch.block?.inferType(expected))
         }
-        return if (expr.elseBranch == null) TyUnit.INSTANCE else getMoreCompleteType(blockTys.filterNotNull())
+        return if (expr.elseBranch == null && expr.parent is RsBlock) {
```

Comment:
Looks like `expr.parent is RsBlock` is always false - if `if` is inside block, its parent will be `RsExprStmt`. Also please add tests when `if` is top-level stmt or tail expr in a function